### PR TITLE
previousVersionFile configured by a string property instead of a function #84

### PR DESCRIPTION
**Pull request by wwilk**
I created this PR to allow others testing it, but it's not thoroughly tested and shouldn't be merged!

Added releasing.previousReleaseVersion.
Removed interface RemoteUrlResolver, two different string properties, that you can set directly on task, are used instead - previousVersionPomUrl and previousVersionSourcesJarUrl

### CODE REVIEW

File: src/main/groovy/org/mockito/release/internal/comparison/BintrayRemoteUrlResolver.java [hunk 1]
```diff
@@ -1,18 +0,0 @@
-package org.mockito.release.internal.comparison;
-
-class BintrayRemoteUrlResolver implements RemoteUrlResolver{
```
Comment (szczepiq) [inline src/main/groovy/org/mockito/release/internal/comparison/BintrayRemoteUrlResolver.java]: Excellent! Simper code!

File: src/main/groovy/org/mockito/release/internal/comparison/PublicationsComparatorTask.java [hunk 2]
```diff
@@ -44,17 +44,19 @@ public boolean isPublicationsEqual() {
         File currentVersionPomPath = pomTask.getDestination();
         File currentVersionSourcesJarPath = sourcesJar.getArchivePath();
 
+        projectName = sourcesJar.getBaseName();
```
Comment (szczepiq) [inline src/main/groovy/org/mockito/release/internal/comparison/PublicationsComparatorTask.java]: the name 'projectName' seems odd, it is the artifact name, isn't it?

File: src/main/groovy/org/mockito/release/internal/comparison/RemoteUrlResolver.java [hunk 3]
```diff
@@ -1,9 +0,0 @@
-package org.mockito.release.internal.comparison;
-
-public interface RemoteUrlResolver {
```
Comment (szczepiq) [inline src/main/groovy/org/mockito/release/internal/comparison/RemoteUrlResolver.java]: I love the smell of deleted code in the evening :)

File: src/main/groovy/org/mockito/release/internal/gradle/BaseJavaLibraryPlugin.java [hunk 4]
```diff
@@ -88,24 +88,7 @@ public void execute(final PublicationsComparatorTask t) {
...
-                        @Override
-                        public void execute(VersioningPlugin versioningPlugin) {
-                        VersionInfo versionInfo = project.getRootProject().getExtensions().getByType(VersionInfo.class);
-
-                        t.setCurrentVersion(project.getVersion().toString());
-                        t.setPreviousVersion(versionInfo.getPreviousVersion());
-                        }
-                    }
-                );
+                t.setPreviousVersion(conf.getPreviousReleaseVersion());
```
Comment (szczepiq) [inline src/main/groovy/org/mockito/release/internal/gradle/BaseJavaLibraryPlugin.java]: How does it feel to replace 20 lines of code with 1-liner? :)

File: src/main/groovy/org/mockito/release/internal/gradle/BaseJavaLibraryPlugin.java [hunk 5]
```diff
@@ -114,8 +97,12 @@ public void execute(VersioningPlugin versioningPlugin) {
                 //The generate pom task is dynamically created by Gradle and we can only access it during execution
                 t.comparePom(POM_TASK);
 
-                t.setProjectGroup(project.getGroup().toString());
-                t.setProjectName(project.getName());
+                LazyConfiguration.lazyConfiguration(t, new Runnable() {
```
Comment (szczepiq) [inline src/main/groovy/org/mockito/release/internal/gradle/BaseJavaLibraryPlugin.java]: Let's use deferred configuration, it's more fitting this case. For more info check the docs on lazy vs deferred configuration.

File: src/main/groovy/org/mockito/release/internal/gradle/ContinuousDeliveryPlugin.java [hunk 6]
```diff
@@ -219,7 +220,7 @@ public void execute(BaseJavaLibraryPlugin p) {
 
                                 //TODO WW, removing comparing publications from the workflow for now
                                 //by commenting out below code
-                                //t.addPublicationsComparator((PublicationsComparatorTask) task);
+                                t.addPublicationsComparator((PublicationsComparatorTask) task);
```
Comment (szczepiq) [inline src/main/groovy/org/mockito/release/internal/gradle/ContinuousDeliveryPlugin.java]: sure? we can iterate on the solution some more and hook it up to the workflow when you're ready :)

File: src/main/groovy/org/mockito/release/internal/gradle/BaseJavaLibraryPlugin.java (hunk 4 above)
Comment (wwilk) [inline src/main/groovy/org/mockito/release/internal/gradle/BaseJavaLibraryPlugin.java]: Awesome!

File: src/main/groovy/org/mockito/release/internal/gradle/ContinuousDeliveryPlugin.java (hunk 6 above)
Comment (wwilk) [inline src/main/groovy/org/mockito/release/internal/gradle/ContinuousDeliveryPlugin.java]: Answer below

**Comment by wwilk** [comment]
@szczepiq 
I tested it with mockito on a few cases:
- adding a System.out statement to one of src/main/java .java files -> release needed (sources jar modification)
- adding a compile time dependency on guava to .gradle file -> release needed (pom modification)
- adding a System.out statement to one of src/test/java .java files -> release NOT needed
- adding an arbitrary dir to .gitignore -> release NOT needed

All tests (except .gitignore) done on mockito-core (root) and mockito-android projects.

Before each test I removed project .gradle dir and did ./gradlew clean assertReleaseNeeded. For comparing sources.jar it's useful to turn on "info" level of logging. 

As soon as this PR is merged I will create a pull request in Mockito with the changes in configuration.

I think it's OK to enable it now. The sooner we start testing it on real products the better, but if you had a minute to see for yourself if it complies with your expectations it would be great!

**Comment by szczepiq** [comment]
I tested with 'example' project and it failed with exception: https://gist.github.com/szczepiq/e13e16392dc66d0cc2e0fbe8ce5c1015

**Comment by szczepiq** [comment]
I tested with Mockito project and found potential issues ([failing comparePublications task](https://gist.github.com/szczepiq/1a2c011a975ed6c574b91ea1936d2720)). I'll unhook from the workflow and merge. There's a lot of good stuff in this PR that I don't want to wait :)

**Pull request merged**